### ISSUE TITLE
Improving performance of TckFile when reading huge MRtrix tck files

### ISSUE THREAD

**Issue by blezek** [nipy/nibabel#995]
Hi,

  First off, thank you all for the great toolkit!  Wanted to point out a performance issue and post a possible solution for others to leverage.

  Using the default `.tck` reader (`TckFile`), 500,000 streamlines takes approximately 22 minutes to read.  Profiling the code reveals most of the reading time is spent on [these lines of array_sequence.py](https://github.com/nipy/nibabel/blob/95e1fbece92b92fc5a5f5c107bd0222438bdb0cc/nibabel/streamlines/array_sequence.py#L237-L238).  Not quite sure what's going on, but reading one streamline at a time thrashes with huge numbers of calls to `numpy.copy`.  Using the code below, the same file is loaded in around 15 seconds.  

  This version made reading faster by reading 512Mb of the 3.5Gb file at a time, then loading all the streamlines into memory at once.  `ArraySequence` performed very well due to [this code](https://github.com/nipy/nibabel/blob/95e1fbece92b92fc5a5f5c107bd0222438bdb0cc/nibabel/streamlines/array_sequence.py#L302-L313).

```python

def load_tck(filename, buffer_size=512):
    """Load a TCK file with a _much_ larger buffer, 512Mb"""
    header = nib.streamlines.tck.TckFile._read_header(filename)
    reader = nib.streamlines.tck.TckFile._read(filename, header, buffer_size=buffer_size)

    start_time = time.time()
    e = [x for x in reader]
    logging.debug(f"read streamline file {filename} in {time.time() - start_time} seconds")
    streamlines = ArraySequence(e)
    tractogram = Tractogram(streamlines)

    # By definition.
    tractogram.affine_to_rasmm = np.eye(4)
    header[Field.VOXEL_TO_RASMM] = np.eye(4)

    return TckFile(tractogram, header=header)
```

**Comment by matthew-brett** [comment]
Ow - that's a huge speedup.  Thanks for reporting.  How low can the buffer size go before you're losing most of the benefit?  @MarcCote - any thoughts?

**Comment by blezek** [comment]
@matthew-brett Testing results seem to indicate it's not the buffer size, 16Mb seems to be as good as anything.

```python

def test_load(filename):
    for buffer_size in [4,8,16,32,64,128,256,512]:
        start = time()
        load_tck(filename, buffer_size=buffer_size)
        diff = time() - start
        logging.info ( f"buffer(Mb): {buffer_size} loaded in {diff}")


test_load ( filename )
INFO:buffer(Mb): 4 loaded in 21.824169158935547
INFO:buffer(Mb): 8 loaded in 15.03981900215149
INFO:buffer(Mb): 16 loaded in 13.3696768283844
INFO:buffer(Mb): 32 loaded in 14.87663221359253
INFO:buffer(Mb): 64 loaded in 13.381511211395264
INFO:buffer(Mb): 128 loaded in 14.74131178855896
INFO:buffer(Mb): 256 loaded in 14.689419746398926
INFO:buffer(Mb): 512 loaded in 15.29773473739624
```
and the `.tck` file info:

```sh
-rw-r--r--  1 mra9161  MFAD\Domain Users   3.4G Feb  3 10:44 streamlines.tck
18: tckinfo 'streamlines.tck'
***********************************
    count:                500000
    init_threshold:       0.100000001
    max_angle:            60
    max_dist:             227.769287
    max_num_seeds:        500000000
    max_num_tracks:       500000
    max_seed_attempts:    1000
    method:               TensorProb
    min_dist:             60
    mrtrix_version:       3.0.2-4-g994abf74
    rk4:                  0
    source:               dti_denoised.mif
    step_size:            0.227769285
    stop_on_all_include:  0
    threshold:            0.100000001
    timestamp:            1612368315.7630646229
    total_count:          1014341
    unidirectional:       0
    ROI:                  mask mask.mif
    ROI:                  seed mask.mif
```

**Comment by matthew-brett** [comment]
Aha - I'm lost now - so what changes made the huge difference to the timing, do you think?    I believe the default buffer size is 4MB, so you timings predict a ~30% speedup from the larger buffer size - right?

**Comment by blezek** [comment]
> Aha - I'm lost now - so what changes made the huge difference to the timing, do you think? I believe the default buffer size is 4MB, so your timings predict a ~30% speedup from the larger buffer size - right?

The biggest contributor to speed up is this code:

```python
    e = [x for x in reader]
    logging.debug(f"read streamline file {filename} in {time.time() - start_time} seconds")
    streamlines = ArraySequence(e)
    tractogram = Tractogram(streamlines)
```

Note, I read all the streamlines immediately.  The normal reader streams them in one at a time, [trying to intelligently cache](https://github.com/nipy/nibabel/blob/95e1fbece92b92fc5a5f5c107bd0222438bdb0cc/nibabel/streamlines/array_sequence.py#L237-L238), but this caching caused huge numbers of `np.copy` calls.  If all the streamlines are already in memory, the load is very quick (slower than `MRTrix`, but very acceptable).

If this [line that extends the streamlines](https://github.com/nipy/nibabel/blob/95e1fbece92b92fc5a5f5c107bd0222438bdb0cc/nibabel/streamlines/array_sequence.py#L135):

```python
        self.extend(iterable)
```

were replaced with

```python
        self.extend( [ x for x in iterable] )
```

you would see the same speed up.  The streamlines would all be read before the call to `self.extend` and the load should go faster.

**Comment by effigies** [comment]
Could you share the benchmark over buffer sizes with the current loader? The idea of using an iterable is so that the max memory usage should be ~ `total + buffer_size`, rather than `2 * total`.

The issue is caused by a change in numpy that restricted 0-copy reshapes. I guess we should revisit this and see if we can get back to a situation where we can reliably avoid copies.

**Comment by blezek** [comment]
@effigies sure, this code:

```python
    for buffer_size in sizes:
        start = time()
        header = nib.streamlines.tck.TckFile._read_header(filename)
        reader = nib.streamlines.tck.TckFile._read(filename, header, buffer_size=buffer_size)
        streamlines = ArraySequence(reader)
        diff = time() - start
        logging.info ( f"buffer(Mb): {buffer_size} loaded in {diff}")
```

`buffer_size` is not exposed to the general reader.

outputs:

```
2021-02-05 10:39:55,337 - root - INFO - buffer(Mb): 4 loaded in 20.610244035720825
2021-02-05 10:40:11,205 - root - INFO - buffer(Mb): 8 loaded in 15.865175247192383
2021-02-05 10:40:27,410 - root - INFO - buffer(Mb): 16 loaded in 16.2043719291687
2021-02-05 10:40:41,325 - root - INFO - buffer(Mb): 32 loaded in 13.915029048919678
2021-02-05 10:40:54,449 - root - INFO - buffer(Mb): 64 loaded in 13.123391151428223
2021-02-05 10:41:08,982 - root - INFO - buffer(Mb): 128 loaded in 14.533066987991333
2021-02-05 10:41:21,472 - root - INFO - buffer(Mb): 256 loaded in 12.489468812942505
2021-02-05 10:41:33,969 - root - INFO - buffer(Mb): 512 loaded in 12.497168064117432

# Existing loader
2021-02-05 11:04:43,840 - root - INFO - buffer(Mb): 4 loaded in 1389.8615169525146
2021-02-05 11:21:37,447 - root - INFO - buffer(Mb): 8 loaded in 1013.5968930721283
2021-02-05 11:38:44,888 - root - INFO - buffer(Mb): 16 loaded in 1027.4412620067596
2021-02-05 11:56:12,701 - root - INFO - buffer(Mb): 32 loaded in 1047.810264825821
2021-02-05 12:13:10,996 - root - INFO - buffer(Mb): 64 loaded in 1018.2923448085785
2021-02-05 12:32:06,154 - root - INFO - buffer(Mb): 128 loaded in 1135.1559188365936

#Killed, because Dan got impatient...
```

**Comment by MarcCote** [comment]
Completely missed this issue. Here's a potential fix: https://github.com/nipy/nibabel/pull/1093.

Benchmarking on a 2.9Gb TCK file (10M streamlines), I got from 300 secs to 24 secs.

As @blezek mentioned, there a room for improvement. If we are willing to trade RAM for speed, we could simply load the entire data from the file (as raw data), then create the ArraySequence from it. In theory, the RAM usage should **_peak_** at around twice the size of the final ArraySequence, then drop down to the size of the ArraySequence.

Note that `[ x for x in iterable]` is exactly what we are trying to avoid with `ArraySequence`. It has been a while since I've checked that, but a list of ndarrays consumes a lot of RAM (probably due to the metadata needed for each ndarray).
